internal/merkle/proof.go

```diff
@@ -0,0 +1,79 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+package merkle
+
+import (
+	"fmt"
+
+	"github.com/cartesi/rollups-node/internal/node/model"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// CreateProofs creates proofs for all the leaves of a binary Merkle tree
+// with the given height. It returns the root hash and the siblings matrix
+// encoded as an array, in bottom-up order.
+//
+// If the number of leaves exceeds the capacity for the given height,
+// an error is returned.
+func CreateProofs(leaves []model.Hash, height uint) (model.Hash, []model.Hash, error) {
+	pristineNode := model.Hash{}
+
+	currentLevel := leaves
+	leafCount := uint(len(leaves))
+	siblings := make([]model.Hash, leafCount*height)
+
+	// for each level in the tree, starting from the leaves
+	for levelIdx := uint(0); levelIdx < height; levelIdx++ {
+		// for each leaf
+		for leafIdx := uint(0); leafIdx < leafCount; leafIdx++ {
+			// calculate its sibling at the current level.
+			// The sibling index is to the left of the parent if leaf index
+			// is odd or to the right if even
```

Comment:
I would expand on this. Feel free to adapt it however you like.

```suggestion
			// Each pair of siblings shares the same parent.
			// For any given parent with index i,
			// the index of its left node is 2*i,
			// and the index of its right node is 2*i+1.
			// Since their indices differ only by the LSB,
			// by flipping the LSB with ^1,
			// we calculate the index of one another.
			// In order to get the index of a parent,
			// you remove the LSB with >>1.
			// For an ancestor of height difference h,
			// you remove the h LSBs with >>h.
```